pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -0,0 +1,1134 @@
+                match walreceiver_state.wal_connection.as_mut() {
+                    Some(wal_connection) => {
+                        let receiver = &mut wal_connection.connection_task.events_receiver;
+                        Some(match receiver.changed().await {
+                            Ok(()) => receiver.borrow().clone(),
+                            Err(_cancellation_error) => TaskEvent::End(Ok(())),
+                        })
+                    }
+                    None => None,
+                }
+            } => {
+                let (connection_update, reset_connection_attempts) = match &wal_connection_update {
+                    TaskEvent::Started => (Some(Utc::now().naive_utc()), true),
+                    TaskEvent::NewEvent(replication_feedback) => (Some(DateTime::<Local>::from(replication_feedback.ps_replytime).naive_utc()), true),
+                    TaskEvent::End(end_result) => {
+                        let should_reset_connection_attempts = match end_result {
+                            Ok(()) => {
+                                debug!("WAL receiving task finished");
+                                true
+                            },
+                            Err(e) => {
+                                warn!("WAL receiving task failed: {e}");
+                                false
+                            },
+                        };
+                        walreceiver_state.wal_connection = None;
+                        (None, should_reset_connection_attempts)
+                    },
+                };
+
+                if let Some(connection_update) = connection_update {
+                    match &mut walreceiver_state.wal_connection {
+                        Some(wal_connection) => {
+                            wal_connection.latest_connection_update = connection_update;
+
+                            let attempts_entry = walreceiver_state.wal_connection_attempts.entry(wal_connection.sk_id).or_insert(0);
+                            if reset_connection_attempts {
+                                *attempts_entry = 0;
+                            } else {
+                                *attempts_entry += 1;
+                            }
+                        },
+                        None => error!("Received connection update for WAL connection that is not active, update: {wal_connection_update:?}"),
+                    }
+                }
+            },
+
+            // If nothing else happens, ensure we are not stuck waiting and still react on etcd changes.
+            // Register the value from the broker or abort the loop, if the subscription got cancelled.
+
+            // XXX: etcd changes are frequent, so don't place it above the slower source events such as WAL receiver one, otherwise
+            // the slow ones might get ignored constantly.
+            broker_update = broker_subscription.value_updates.recv() => {
```

Comment:
I realized with this patch watch is never canceled, so task is leaked if connection manager is canceled. Perhaps it would be ok to add cancel to `BrokerSubscription` impl Drop. Hmm, on the second thought this leak is very short living as it will fail sending to closed channel.

Reply:
I think it's ok to `imp Drop`, thank you for spotting.

Indeed, we're very much guarded buy all these shared channel ends being dropped as signals to stop working.
IMO, it's a bit dangerous since on the top level, where we register new timelines with endless tasks, we spawn channel-less endless tasks to poll etcd and if they panic, we won't even know that now, since we never call `handle.await` before the shutdown.

The only solution I see now is to have a separate loop (that also may panic) to check for such tasks constantly, restarting the abruptly stopped ones, but that requires way more changes that I'm ready to do now.